File: haskell/src/Algorithms/Lloyd/Sequential.hs
{-# LANGUAGE ViewPatterns #-}

module Algorithms.Lloyd.Sequential (
  Point(..),
  Cluster(..),
  ExpectDivergent(..),
  kmeans,
  PointSum(..),
  makeNewClusters,
  assign,
  assignPS,
  step
)where

import Prelude hiding (zipWith, map, foldr, replicate, length, zip, head)
import Control.Applicative ((<$>))
import Control.Monad (guard)
import Data.Foldable (Foldable(foldr))
import Data.Function (on)
import Data.Functor.Extras ((..:),(...:))
import Data.Metric (Metric(..))
import Data.Semigroup (Semigroup(..))
import Data.Vector (Vector(..), toList, fromList, create, zip, zipWith, map, replicate, minimumBy, length, head, forM_)
import qualified Data.Vector.Mutable as MV (replicate, read, write)

data Point = Point
  { point :: Vector Double
  } deriving (Eq, Show)

pmap :: (Vector Double -> Vector Double) -> Point -> Point
pmap f = Point . f . point

useMetric :: Metric a => (Vector Double -> a) -> Point -> Point -> Double
useMetric metric = distance `on` (metric . point)

instance Semigroup Point where
  (<>) = Point ..: zipWith (+) `on` point

data Cluster = Cluster
  { identifier :: Int
  , centroid   :: Point
  } deriving (Show)

instance Eq Cluster where
  (==) = (==) `on` centroid

data PointSum = PointSum Int Point deriving (Show)

instance Semigroup PointSum where
  PointSum c0 p0 <> PointSum c1 p1 = PointSum (c0+c1) (p0<>p1)

emptyPointSum :: Int -> PointSum
emptyPointSum length = PointSum 0 . Point $ replicate length 0

toCluster :: Int -> PointSum -> Cluster
toCluster cid (PointSum count point) = Cluster
  { identifier = cid
  , centroid   = map (//count) `pmap` point
  }

(//) :: Double -> Int -> Double
x // y = x / fromIntegral y

closestCluster :: Metric a => (Vector Double -> a) -> Vector Cluster -> Point -> Cluster
closestCluster (useMetric -> d) clusters point = fst . minimumBy (compare `on` snd) $ do
  cluster <- clusters
  return (cluster, point `d` centroid cluster)

-- Type system can't handle `map fromList . create` here; ($) is magical according to John Hughes.
assign :: Metric a => (Vector Double -> a) -> Vector Cluster -> Vector Point -> Vector (Vector Point)
assign metric clusters points = map fromList $ create $ do
  vector <- MV.replicate (length clusters) []
  points `forM_` \point -> do
    let cluster  = closestCluster metric clusters point
        position = identifier cluster
    points' <- MV.read vector position
    MV.write vector position $ point : points'
  return vector

assignPS :: Metric a => (Vector Double -> a) -> Vector Cluster -> Vector Point -> Vector PointSum
assignPS metric clusters points = reduce <$> assign metric clusters points
  where reduce  = foldr (<>) (emptyPointSum length') . fmap (PointSum 1)
        length' = length . point $ head points

makeNewClusters :: Vector PointSum -> Vector Cluster
makeNewClusters vector = do
  (pointSum, index) <- zip vector $ fromList [0..length vector]
  return $ toCluster index pointSum

step :: Metric a => (Vector Double -> a) -> Vector Cluster -> Vector Point -> Vector Cluster
step = makeNewClusters ...: assignPS

newtype ExpectDivergent = ExpectDivergent { expectDivergent :: Int }

computeClusters :: Metric a => ExpectDivergent -> (Vector Double -> a) -> Vector Point -> Vector Cluster -> Vector Cluster
computeClusters (expectDivergent -> expectDivergent) metric = computeClusters' expectDivergent metric 0

computeClusters' :: Metric a => Int -> (Vector Double -> a) -> Int -> Vector Point -> Vector Cluster -> Vector Cluster
computeClusters' expectDivergent metric iterations points clusters
  | iterations >= expectDivergent = clusters
  | clusters' == clusters         = clusters
  | otherwise                     = computeClusters' expectDivergent metric (succ iterations) points clusters'
  where clusters' = step metric clusters points

kmeans :: Metric a => ExpectDivergent -> (Vector Double -> a) -> Vector Point -> Vector Cluster -> Vector (Vector Point)
kmeans expectDivergent metric points initial = assign metric clusters points
  where clusters = computeClusters expectDivergent metric points initial

File: scala/benchmarks/src/main/scala/programs/Lloyd/Sequential.scala
package Lloyd

/*
This is almost a direct translation of Algorithms.Lloyd.Sequential from a
modified version of the Haskell package kmeans-par.
*/

import scala.collection.parallel._
import scala.collection.mutable
import vector._


package object Sequential {

type Point = Vector[Double]
type Metric[T] = (T, T) => Double

// Semigroup instance
def combinePoints(a: Point, b: Point): Point = {
  val add = (a: Double, b: Double) => a + b
  a.zip(b).map(add.tupled)
}

case class Cluster(
  identifier: Int,
  centroid: Point,
)

def eqClusters(as: Vector[Cluster], bs: Vector[Cluster]): Boolean = {
  val eq = (a: Cluster, b: Cluster) => a.centroid == b.centroid
  as.length == bs.length && as.zip(bs).forall(eq.tupled)
}

case class PointSum(
  count: Int,
  sum: Point,
)

// Semigroup instance
def combinePointSums(a: PointSum, b: PointSum): PointSum =
  PointSum(a.count+b.count, combinePoints(a.sum, b.sum))

def emptyPointSum(length: Int): PointSum =
  PointSum(0, Vector.tabulate(length)(_ => 0.0))

def toCluster(pointSum: PointSum, cid: Int): Cluster =
  Cluster(cid, pointSum.sum.map(_ / pointSum.count))

def closestCluster(
  metric: Metric[Point],
  clusters: Vector[Cluster],
  point: Point
): Cluster = {
  val clustersAndDistances = clusters.map(c => (c, metric(point, c.centroid)))
  clustersAndDistances.minBy(_._2)._1 // compare on snd; return fst
}

def assign(
  metric: Metric[Point],
  clusters: Vector[Cluster],
  points: Vector[Point],
): Vector[Vector[Point]] = {
  val array: Array[List[Point]] = Array.tabulate(clusters.length)(_ => Nil)
  points.foreach(point => {
    val cluster = closestCluster(metric, clusters, point)
    val position = cluster.identifier
    array(position) = point :: array(position)
  })
  array.map(_.toVector).toVector
}

def assignPS(
  metric: Metric[Point],
  clusters: Vector[Cluster],
  points: Vector[Point],
): Vector[PointSum] = {
  val z = emptyPointSum(points.head.length)
  val reduce = (_: Vector[Point]).map(PointSum(1, _)).fold(z)(combinePointSums)
  assign(metric, clusters, points).map(reduce)
}

def makeNewClusters(vector: Vector[PointSum]): Vector[Cluster] =
  // MODIFIED: map instead of flatMap and singleton
  vector.zipWithIndex.map((toCluster _).tupled)

def step(
  metric: Metric[Point],
  clusters: Vector[Cluster],
  points: Vector[Point],
): Vector[Cluster] =
  makeNewClusters(assignPS(metric, clusters, points))

type ExpectDivergent = Int

def computeClusters(
  expectDivergent: ExpectDivergent,
  metric: Metric[Point],
  points: Vector[Point],
  clusters: Vector[Cluster],
): Vector[Cluster] =
  computeClusters_(expectDivergent, metric, 0, points, clusters)

def computeClusters_(
  expectDivergent: ExpectDivergent,
  metric: Metric[Point],
  iterations: Int,
  points: Vector[Point],
  clusters: Vector[Cluster],  
): Vector[Cluster] = {
  val newClusters = step(metric, clusters, points)
  Unit match {
    case _ if iterations >= expectDivergent     => clusters
    case _ if eqClusters(newClusters, clusters) => clusters
    case _                                      =>
      computeClusters_(expectDivergent, metric, iterations+1, points, newClusters)
  }
}

def kmeans(
  expectDivergent: ExpectDivergent,
  metric: Metric[Point],
  points: Vector[Point],
  initial: Vector[Cluster],
): Vector[Vector[Point]] = {
  val clusters = computeClusters(expectDivergent, metric, points, initial)
  assign(metric, clusters, points)
}


// end of package object
}
